python/lsst/daf/butler/datastores/fileDatastore.py

```diff
@@ -1070,11 +1067,20 @@ def _removeFileExists(uri: ButlerURI) -> None:
                 uri.write(serializedDataset, overwrite=True)
                 log.debug("Successfully wrote bytes directly to %s", uri)
 
+                # It may be that this dataset should be cached locally
```

Comment:
It's not serializing the in-memory dataset twice -- the bytes are just bytes that are written twice. Rewriting it so that we always write to the local file first if the intent is to be cached seems reasonable (I hadn't really intended to cache these at all initially -- the tests all failed without because the tests all use very small datasets that can bypass the file writing stage -- hence the convoluted logic I ended up with). 

There is a related problem that if the temp directory is on a different file system to the cache directory we will have to (effectively) read all the bytes back into memory for the "move" since `os.rename` won't work.

Reply:
Thanks. The rewrite is much clearer.